cirq-google/cirq_google/engine/engine_sampler.py

```diff
@@ -138,3 +138,118 @@ def get_engine_sampler(
         )
     gate_set = gate_sets.NAMED_GATESETS[gate_set_name]
     return engine.get_engine(project_id).sampler(processor_id=processor_id, gate_set=gate_set)
+
+
+def get_device_sampler(
+    project_id: Optional[str] = None, processor_id: Optional[str] = None, get_simulator: bool = True
+) -> Union[
+    bool,
+    Tuple[
+        Tuple[Union[cirq.Device], int],
+        Union['cirq_google.PhasedFSimEngineSimulator', 'cirq_google.QuantumEngineSampler'],
+    ],
+]:
+    """Authenticates on Google Cloud, can return a Device and Simulator.
+
+    This uses the environment variable GOOGLE_CLOUD_PROJECT for the Engine
+    project_id, unless set explicitly.
+
+    Args:
+        project_id: Optional explicit Google Cloud project id. Otherwise,
+            this defaults to the environment variable GOOGLE_CLOUD_PROJECT.
+            By using an environment variable, you can avoid hard-coding
+            personal project IDs in shared code.
+        processor_id: Engine processor ID (from Cloud console or
+            ``Engine.list_processors``).
+        get_simulator: Boolean value indicating whether to return a device
+            and simulator after authentication or not.
+
+    Returns:
```

Comment:
It's generally a code smell to have the return _type_ change with different arguments. Consider always returning bool, device, sampler.

If one of those truly doesn't make sense for a given input, set it to `None`. For this specific case: if `get_simulator` is True, I'd expect the sampler to be `cirq.Simulator()` which implements the `Sampler` interface so all is well

Reply:
sorry, I originall read this backwards. If `get_simulator` is False, I'd expect the function to return an EngineSampler and a real `Device`

Reply:
Ok I think I misunderstood again. We use the word "simulator" to refer to a simulated quantum processor. If the user specifies a valid processor id and project id and can authenticate, they will receive a sampler backed by a real quantum processor, i.e. not a simulator at all. So I think the variable named `get_simulator` is misnamed. It also may be redundant. You can always just return one. The notebook can ignore it if it doesn't want one. Instantiating a sampler and device is not an expensive operation

Reply:
I removed the get simulator argument, things are a lot cleaner now, thanks for the advice. `get_simulator` is altogether removed.
I have fix the tests, but by suppressing coverage of the auth functions.